File: platforms/android/runtime/core/src/main/java/io/nativeblocks/runtime/api/NativeblocksManager.kt
package io.nativeblocks.runtime.api

import android.content.Context
import androidx.compose.runtime.Composable
import io.nativeblocks.runtime.api.provider.action.INativeAction
import io.nativeblocks.runtime.api.provider.action.INativeActionContractor
import io.nativeblocks.runtime.api.provider.action.NativeActionProviderRegistry
import io.nativeblocks.runtime.api.provider.block.BlockProps
import io.nativeblocks.runtime.api.provider.block.NativeBlockProviderRegistry
import io.nativeblocks.runtime.api.provider.kit.Kit
import io.nativeblocks.runtime.api.provider.logger.INativeLogger
import io.nativeblocks.runtime.api.provider.model.NativeScaffoldModel
import io.nativeblocks.runtime.api.provider.type.INativeType
import io.nativeblocks.runtime.api.provider.type.NativeTypeProvider
import io.nativeblocks.runtime.api.provider.type.NativeTypeProviderRegistry
import io.nativeblocks.runtime.di.NativeCoreSDKInjector
import io.nativeblocks.runtime.experiment.ExperimentUseCase
import io.nativeblocks.runtime.ffi.NativeRuntimeClientManager
import io.nativeblocks.runtime.ffi.disposeInstance
import io.nativeblocks.runtime.ffi.provideLogger
import io.nativeblocks.runtime.ffi.removeLogger
import io.nativeblocks.runtime.localization.LocalizationUseCase
import io.nativeblocks.runtime.logger.LoggerAdapter
import io.nativeblocks.runtime.scaffold.toHost
import io.nativeblocks.runtime.util.isValidInstanceName
import org.koin.core.qualifier.named
import kotlin.reflect.KClass

/**
 * Manages the lifecycle and configuration of the Nativeblocks framework.
 */
class NativeblocksManager internal constructor(
    private val name: String,
    private val context: Context,
    private val edition: NativeblocksEdition
) {

    companion object {
        /**
         * Singletons instance of the NativeblocksManager.
         */
        private val instanceRegistry = mutableMapOf<String, NativeblocksManager>()

        private val lock = Any()

        /**
         * Initializes the Nativeblocks framework.
         * @param applicationContext The application context.
         * @param edition The edition of Nativeblocks to initialize.
         */
        fun initialize(
            name: String = "default",
            applicationContext: Context,
            edition: NativeblocksEdition
        ): NativeblocksManager {
            synchronized(lock) {
                if (name.isValidInstanceName().not()) {
                    throw IllegalArgumentException("Please make sure the '$name' contains valid characters: A–Z, a–z, 0–9, _ or -")
                }
                if (!instanceRegistry.containsKey(name)) {
                    val manager = NativeblocksManager(name, applicationContext, edition)
                    instanceRegistry[name] = manager
                }
                return instanceRegistry[name]!!
            }
        }

        /**
         * Retrieves the singleton instance of the NativeblocksManager.
         * @return The NativeblocksManager instance.
         */
        fun getInstance(name: String = "default"): NativeblocksManager {
            return instanceRegistry[name]
                ?: throw IllegalStateException("NativeblocksManager '$name' has not been initialized.")
        }

        /**
         * Check whether the instance has been initialized or not
         * @return Boolean
         */
        fun isInitialized(name: String = "default"): Boolean {
            return instanceRegistry.containsKey(name)
        }
    }

    init {
        NativeCoreSDKInjector.init(this.name, context, edition)
    }

    private fun getKoin() = NativeCoreSDKInjector.get(this.name).koin
    private val typeProvider: NativeTypeProvider = NativeTypeProviderRegistry.getOrCreate(this.name)

    private val blockProvider get() = NativeBlockProviderRegistry.getOrCreate(this.name)
    private val actionProvider get() = NativeActionProviderRegistry.getOrCreate(this.name)
    private val actionContractors = mutableListOf<INativeActionContractor>()
    private val loggerTypes = mutableSetOf<String>()
    private val kits = mutableListOf<Kit>()

    internal fun providedActionContractors(): List<INativeActionContractor> = actionContractors.toList()

    /**
     * Provides a block implementation.
     * @param blockType The type of the block.
     * @param block The block implementation to register.
     * @return The NativeblocksManager instance for chaining.
     */
    fun provideBlock(
        blockType: String,
        block: @Composable (blockProps: BlockProps) -> Unit
    ): NativeblocksManager {
        blockProvider.provideBlock(blockType, block)
        return this
    }

    /**
     * Provides a fallback block to be used when a requested block keyType is not found.
     * @param block A composable lambda that receives the missing block's keyType and key.
     * @return The NativeblocksManager instance for chaining.
     */
    fun provideFallbackBlock(block: @Composable (keyType: String, key: String) -> Unit): NativeblocksManager {
        blockProvider.onFallbackBlock(block)
        return this
    }

    /**
     * Provides an action implementation.
     * @param actionType The type of the action.
     * @param action The action implementation to register.
     * @return The NativeblocksManager instance for chaining.
     */
    fun provideAction(actionType: String, action: INativeAction): NativeblocksManager {
        actionProvider.provideAction(actionType, action)
        return this
    }

    /**
     * Provides an action contractor.
     * @param actionContractor The action contractor to provide.
     * @return The NativeblocksManager instance for chaining.
     */
    fun provideActionContractor(actionContractor: INativeActionContractor): NativeblocksManager {
        actionContractors.add(actionContractor)
        return this
    }

    /**
     * Provides a fallback action to be used when a requested action keyType is not found.
     * @param block A lambda that receives the missing action's keyType and name.
     * @return The NativeblocksManager instance for chaining.
     */
    fun provideFallbackAction(block: (keyType: String, name: String) -> Unit): NativeblocksManager {
        actionProvider.onFallbackAction(block)
        return this
    }

    /**
     * Provides an event logger.
     * @param loggerType The type of the logger.
     * @param logger The logger implementation to provide.
     * @return The NativeblocksManager instance for chaining.
     */
    fun provideEventLogger(loggerType: String, logger: INativeLogger): NativeblocksManager {
        runCatching {
            provideLogger(this.name, loggerType, LoggerAdapter(logger))
            loggerTypes.add(loggerType)
        }
        return this
    }

    /**
     * Provides a kit.
     * @param kit The kit to provide.
     * @return The NativeblocksManager instance for chaining.
     */
    fun provideKit(kit: Kit): NativeblocksManager {
        runCatching {
            kit.attach(this.context, this.name, this.edition)
            kits.add(kit)
        }
        return this
    }

    /**
     * Synchronizes the specified frame with the server.
     * @param route The route of the frame to synchronize.
     * @return The NativeblocksManager instance for chaining.
     */
    suspend fun syncFrame(route: String): NativeblocksManager {
        val runtimeClient: NativeRuntimeClientManager by getKoin().inject(named(this.name))
        runCatching { runtimeClient.frameClient.syncFrame(route, emptyMap()) }
        return this
    }

    /**
     * clear all frames from cache.
     * @return The NativeblocksManager instance for chaining.
     */
    suspend fun clearAllFrames(): NativeblocksManager {
        val runtimeClient: NativeRuntimeClientManager by getKoin().inject(named(this.name))
        runCatching { runtimeClient.frameClient.clearAll(emptyList()) }
        return this
    }

    /**
     * clear frame from cache.
     * @param route The route of the frame to clear.
     * @return The NativeblocksManager instance for chaining.
     */
    suspend fun clearFrame(route: String): NativeblocksManager {
        val runtimeClient: NativeRuntimeClientManager by getKoin().inject(named(this.name))
        runCatching { runtimeClient.frameClient.clear(route) }
        return this
    }

    /**
     * Throws away the frame kept under [key], so the next visit starts fresh.
     * @param key The key given to [NativeblocksFrameState.Stateful].
     */
    fun clearFrameState(key: String): NativeblocksManager {
        val runtimeClient: NativeRuntimeClientManager by getKoin().inject(named(this.name))
        runtimeClient.frameClient.clearFrameState(key)
        return this
    }

    /**
     * Throws away every kept frame.
     */
    fun clearAllFrameStates(): NativeblocksManager {
        val runtimeClient: NativeRuntimeClientManager by getKoin().inject(named(this.name))
        runtimeClient.frameClient.clearAllFrameStates()
        return this
    }

    /**
     * Sets the language for the frames.
     * @param languageCode The ISO language code to set as the current language (e.g. "EN", "FA")
     * @return The NativeblocksManager instance for chaining.
     */
    fun setLanguage(languageCode: String): NativeblocksManager {
        val localizationUseCase: LocalizationUseCase by getKoin().inject(named(this.name))
        localizationUseCase.setLocalization(languageCode = languageCode)
        return this
    }

    /**
     * Translate a key in provided language.
     * @param key The translation key
     * @return The translated value or null.
     */
    fun translate(key: String): String? {
        val localizationUseCase: LocalizationUseCase by getKoin().inject(named(this.name))
        return localizationUseCase.translate(key).getOrNull()
    }

    /**
     * Retrieves the scaffold model for the current configuration.
     */
    suspend fun getScaffold(): Result<NativeScaffoldModel> {
        val runtimeClient: NativeRuntimeClientManager by getKoin().inject(named(this.name))
        return runCatching { runtimeClient.scaffoldClient.getScaffold().toHost() }
    }

    /**
     * Retrieves the experiment value for the provided key as type [T].
     *
     * Supports the following types:
     *  - STRING: String
     *  - NUMBER: Int, Float, Double, Long
     *  - BOOLEAN: Boolean
     *  - JSON: String (as raw JSON)
     *
     * @param key The identifier for the experiment value.
     * @param defaultValue The default value to use if key not found or conversion fails.
     * @param cacheTTL Cache time-to-live in milliseconds. Use 0 for no caching (always fetch fresh).
     *                 Default is 24 hours (86,400,000 ms).
     * @return The experiment value as type [T] or [defaultValue] if unavailable or conversion fails.
     *
     * @sample
     * ```
     * // Use default cache (24 hours)
     * val showNewUI = getExperiment("show_new_ui", false)
     *
     * // No cache - always fetch fresh
     * val criticalFlag = getExperiment("payment_enabled", false, 0L)
     *
     * // Custom cache duration (10 minutes)
     * val theme = getExperiment("theme_color", "#FF0000", 10 * 60 * 1000L)
     * ```
     */
    @Suppress("UNCHECKED_CAST")
    suspend fun <T : Any> getExperiment(key: String, defaultValue: T, cacheTTL: Long = 24 * 60 * 60 * 1000): T {
        val experimentUseCase: ExperimentUseCase by getKoin().inject(named(this.name))
        val result = experimentUseCase.get(key, cacheTTL) ?: return defaultValue
        val (value, variableType) = result

        val requestedType = when (defaultValue) {
            is String -> if (variableType == "JSON") "JSON" else "STRING"
            is Int, is Float, is Double, is Long -> "NUMBER"
            is Boolean -> "BOOLEAN"
            else -> null
        }
        if (requestedType == null || variableType != requestedType) {
            return defaultValue
        }

        return try {
            when (variableType) {
                "STRING" -> value as T
                "NUMBER" -> when (defaultValue) {
                    is Int -> value.toIntOrNull() as? T ?: defaultValue
                    is Float -> value.toFloatOrNull() as? T ?: defaultValue
                    is Double -> value.toDoubleOrNull() as? T ?: defaultValue
                    is Long -> value.toLongOrNull() as? T ?: defaultValue
                    else -> defaultValue
                }
                "BOOLEAN" -> when (defaultValue) {
                    is Boolean -> value.toBooleanStrictOrNull() as? T ?: defaultValue
                    else -> defaultValue
                }
                "JSON" -> value as T
                else -> defaultValue
            }
        } catch (_: Exception) {
            defaultValue
        }
    }

    /**
     * Sets global parameters for the frames.
     * @param @parameters Vararg of pairs representing global parameters, where each pair consists of a key and a value.
     * @return The NativeblocksManager instance for chaining.
     */
    fun setGlobalParameters(vararg parameters: Pair<String, String>): NativeblocksManager {
        return setGlobalParameters(parameters.toMap())
    }

    /**
     * Sets global parameters for the frames.
     * @param @parameters Vararg of pairs representing global parameters, where each pair consists of a key and a value.
     * @return The NativeblocksManager instance for chaining.
     */
    fun setGlobalParameters(parameters: Map<String, String>): NativeblocksManager {
        val runtimeClient: NativeRuntimeClientManager by getKoin().inject(named(this.name))
        runCatching { runtimeClient.globalParameterClient.set(parameters) }
        return this
    }

    /**
     * Provides a type converter for a given type within the NativeblocksManager.
     *
     * @param T The type for which the converter is being provided.
     * @param type The `KClass` of the type.
     * @param converter The implementation of `INativeType` to handle conversion for the type.
     * @return The instance of `NativeblocksManager` to allow method chaining.
     */
    fun <T : Any> provideTypeConverter(
        type: KClass<T>,
        converter: INativeType<T>
    ): NativeblocksManager {
        typeProvider.provideTypeConverter(type, converter)
        return this
    }

    /**
     * Retrieves the type converter for a given type from the NativeblocksManager.
     *
     * @param T The type for which the converter is being retrieved.
     * @param type The `KClass` of the type.
     * @return The `INativeType` implementation for the requested type.
     * @throws NullPointerException If no converter has been provided for the requested type.
     */
    fun <T : Any> getTypeConverter(type: KClass<T>): INativeType<T> {
        return typeProvider.getTypeConverter(type)
    }

    /**
     * Cleans up resources and destroys the NativeblocksManager instance.
     */
    fun destroy() {
        kits.forEach { runCatching { it.detach(this.name) } }
        kits.clear()
        loggerTypes.forEach { runCatching { removeLogger(this.name, it) } }
        loggerTypes.clear()
        runCatching {
            val runtimeClient: NativeRuntimeClientManager by getKoin().inject(named(this.name))
            runtimeClient.close()
        }
        NativeCoreSDKInjector.destroy(this.name)
        runCatching { disposeInstance(this.name) }
        instanceRegistry.remove(this.name)
        NativeBlockProviderRegistry.remove(this.name)
        NativeActionProviderRegistry.remove(this.name)
        NativeTypeProviderRegistry.remove(this.name)
        actionContractors.clear()
    }

}

File: platforms/ios/runtime/Sources/NativeblocksRuntime/api/NativeblocksManager.swift
import Foundation
import NativeblocksRuntimeFFI
import SwiftUI

/// Manages the lifecycle and configuration of the Nativeblocks framework.
/// The `NativeblocksManager` is used to initialize the SDK, register blocks, actions and loggers,
/// and to reach the runtime for frames, localization and experiments.
public class NativeblocksManager {

    private static var instanceRegistry: [String: NativeblocksManager] = [:]
    private static let lock = NSLock()

    private let name: String
    private let edition: NativeblocksEdition
    private let typeProvider: NativeTypeProvider
    private var loggerTypes = Set<String>()
    private var actionContractors = [any INativeActionContractor]()
    private var kits = [any Kit]()

    internal func providedActionContractors() -> [any INativeActionContractor] {
        return actionContractors
    }

    /// Private initializer to enforce singleton usage.
    private init(name: String, edition: NativeblocksEdition) {
        self.name = name
        self.edition = edition
        self.typeProvider = NativeTypeProviderRegistry.getOrCreate(name)
        NativeCoreSDKInjector.initialize(name: name, edition: edition)
    }

    private var injector: NativeCoreSDKInjector {
        return NativeCoreSDKInjector.get(name: name)
    }

    private var blockProvider: NativeBlockProvider {
        return NativeBlockProviderRegistry.getOrCreate(name)
    }

    private var actionProvider: NativeActionProvider {
        return NativeActionProviderRegistry.getOrCreate(name)
    }

    private var modifierProvider: NativeModifierProvider {
        return NativeModifierProviderRegistry.getOrCreate(name)
    }

    /// Initializes the `NativeblocksManager` with the specified edition.
    /// - Parameters:
    ///   - name: The instance name. Valid characters: A–Z, a–z, 0–9, _ or -.
    ///   - edition: The `NativeblocksEdition` that determines the configuration type, such as cloud or community.
    @discardableResult
    public static func initialize(name: String = "default", edition: NativeblocksEdition) -> NativeblocksManager {
        lock.lock()
        defer { lock.unlock() }

        if !name.isValidInstanceName() {
            fatalError("Please make sure the '\(name)' contains valid characters: A–Z, a–z, 0–9, _ or -")
        }
        if let existing = instanceRegistry[name] {
            return existing
        }
        let manager = NativeblocksManager(name: name, edition: edition)
        instanceRegistry[name] = manager
        return manager
    }

    /// Check whether the instance has been initialized or not
    /// - Returns: Boolean.
    public static func isInitialized(name: String = "default") -> Bool {
        lock.lock()
        defer { lock.unlock() }
        return instanceRegistry[name] != nil
    }

    /// Gets the shared instance of the `NativeblocksManager`.
    /// - Returns: The shared `NativeblocksManager` instance.
    public static func getInstance(name: String = "default") -> NativeblocksManager {
        lock.lock()
        defer { lock.unlock() }
        guard let instance = instanceRegistry[name] else {
            fatalError("NativeblocksManager '\(name)' has not been initialized.")
        }
        return instance
    }

    /// Provides a block implementation.
    /// - Parameters:
    ///   - blockType: The type of the block.
    ///   - block: The view builder rendering the block.
    @discardableResult
    public func provideBlock(blockType: String, block: NativeBlock) -> NativeblocksManager {
        blockProvider.provideBlock(blockType: blockType, block: block)
        return self
    }

    /// Provides a modifier implementation.
    /// - Parameters:
    ///   - modifierType: The type of the modifier.
    ///   - modifier: A closure applying the modifier to the content it decorates.
    /// - Returns: The current NativeblocksManager instance.
    @discardableResult
    public func provideModifier(
        modifierType: String,
        modifier: @escaping (AnyView, ModifierContext) -> AnyView
    ) -> NativeblocksManager {
        modifierProvider.provideModifier(modifierType: modifierType, modifier: modifier)
        return self
    }

    /// Provides a fallback block to be used when a requested block keyType is not found.
    /// This block will be displayed in place of unsupported or unrecognized blocks.
    ///
    /// - Parameter block: A view builder receiving the missing block's keyType and key.
    @discardableResult
    public func provideFallbackBlock(block: @escaping (String, String) -> any View) -> NativeblocksManager {
        blockProvider.onFallbackBlock(block: block)
        return self
    }

    /// Provides an action implementation.
    /// - Parameters:
    ///   - actionType: The type of the action.
    ///   - action: The action instance implementing the `INativeAction` protocol.
    @discardableResult
    public func provideAction(actionType: String, action: any INativeAction) -> NativeblocksManager {
        actionProvider.provideAction(actionType: actionType, action: action)
        return self
    }

    /// Provides an action contractor.
    /// - Parameter actionContractor: The action contractor to provide.
    @discardableResult
    public func provideActionContractor(_ actionContractor: any INativeActionContractor) -> NativeblocksManager {
        actionContractors.append(actionContractor)
        actionProvider.provideActionContractor(actionContractor)
        return self
    }

    /// Provides a fallback action to be used when a requested action keyType is not found.
    /// This action will be executed in place of unsupported or unrecognized actions.
    ///
    /// - Parameter block: A closure receiving the missing action's keyType and name.
    @discardableResult
    public func provideFallbackAction(block: @escaping (String, String) -> Void) -> NativeblocksManager {
        actionProvider.onFallbackAction(block: block)
        return self
    }

    /// Provides an event logger for the specified logger type.
    /// The logger is registered with the runtime, so it receives every event the engine emits.
    /// - Parameters:
    ///   - loggerType: The key type associated with the logger.
    ///   - logger: The logger instance implementing the `INativeLogger` protocol.
    @discardableResult
    public func provideEventLogger(loggerType: String, logger: any INativeLogger) -> NativeblocksManager {
        provideLogger(instanceName: name, loggerType: loggerType, logger: LoggerAdapter(delegate: logger))
        loggerTypes.insert(loggerType)
        return self
    }

    /// Provides a kit.
    /// - Parameter kit: The kit to provide.
    @discardableResult
    public func provideKit(_ kit: any Kit) -> NativeblocksManager {
        kit.attach(instanceName: name, edition: edition)
        kits.append(kit)
        return self
    }

    /// Registers a type converter for a specific type.
    /// - Parameters:
    ///   - type: The type for which the converter is being registered.
    ///   - converter: The converter that implements `INativeType` for the given type.
    @discardableResult
    public func provideTypeConverter<T>(_ type: T.Type, converter: INativeType<T>) -> NativeblocksManager {
        typeProvider.provideTypeConverter(type, converter: converter)
        return self
    }

    /// Retrieves the type converter for a specific type.
    /// - Parameter type: The type for which the converter is being retrieved.
    /// - Returns: The converter that implements `INativeType` for the given type.
    public func getTypeConverter<T>(_ type: T.Type) -> INativeType<T> {
        return typeProvider.getTypeConverter(type)
    }

    /// Synchronizes the frame for the specified route asynchronously.
    /// - Parameter route: The route of the frame to synchronize.
    @discardableResult
    public func syncFrame(route: String) async -> NativeblocksManager {
        try? await injector.runtimeClient.frameClient.syncFrame(route: route, parameters: [:])
        return self
    }

    /// Clear all frames from the cache.
    /// - Returns: The `NativeblocksManager` instance for chaining.
    @discardableResult
    public func clearAllFrames() async -> NativeblocksManager {
        try? await injector.runtimeClient.frameClient.clearAll(routes: [])
        return self
    }

    /// Clear a specific frame from the cache.
    /// - Parameter route: The route of the frame to clear.
    /// - Returns: The `NativeblocksManager` instance for chaining.
    @discardableResult
    public func clearFrame(route: String) async -> NativeblocksManager {
        try? await injector.runtimeClient.frameClient.clear(route: route)
        return self
    }

    /// Throws away the frame kept under `key`, so the next visit starts fresh.
    /// - Parameter key: The key given to `NativeblocksFrameState.stateful`.
    @discardableResult
    public func clearFrameState(key: String) -> NativeblocksManager {
        injector.runtimeClient.frameClient.clearFrameState(stateKey: key)
        return self
    }

    /// Throws away every kept frame.
    @discardableResult
    public func clearAllFrameStates() -> NativeblocksManager {
        injector.runtimeClient.frameClient.clearAllFrameStates()
        return self
    }

    /// Retrieves the scaffold for the frames asynchronously.
    /// - Returns: The `NativeScaffoldModel` on success, or the underlying error on failure.
    public func getScaffold() async -> Result<NativeScaffoldModel, Error> {
        do {
            return .success(try await injector.runtimeClient.scaffoldClient.getScaffold().toHost())
        } catch {
            return .failure(error)
        }
    }

    /// Sets the language for the frames.
    /// - Parameter languageCode: The ISO 639-1 language code in uppercase (e.g., "EN" for English, "FR" for French).
    /// - Returns: The current instance of `NativeblocksManager` for method chaining.
    @discardableResult
    public func setLanguage(languageCode: String) -> NativeblocksManager {
        injector.localizationUseCase.setLocalization(languageCode: languageCode)
        return self
    }

    /// Translates a key in the currently selected language.
    /// - Parameter key: The translation key.
    /// - Returns: The translated value, or `nil` when the key is unknown.
    public func translate(key: String) -> String? {
        return injector.localizationUseCase.translate(key: key)
    }

    /// Sets global parameters for the Nativeblocks instance.
    /// - Parameter parameters: A variadic list of key–value pairs representing global parameters.
    /// - Returns: The current instance of `NativeblocksManager` for method chaining.
    @discardableResult
    public func setGlobalParameters(_ parameters: (String, String)...) -> NativeblocksManager {
        var internalParams = [String: String]()
        for (key, value) in parameters {
            internalParams[key] = value
        }
        return setGlobalParameters(internalParams)
    }

    /// Sets global parameters for the Nativeblocks instance.
    /// - Parameter parameters: A dictionary of key–value pairs representing global parameters.
    /// - Returns: The current instance of `NativeblocksManager` for method chaining.
    @discardableResult
    public func setGlobalParameters(_ parameters: [String: String]) -> NativeblocksManager {
        injector.runtimeClient.globalParameterClient.set(parameters: parameters)
        return self
    }

    /// Fetches an experiment value for the given key with type-safe conversion.
    /// The method validates the server-returned variable type against the expected type and returns
    /// the default value on any error or type mismatch.
    ///
    /// - Parameters:
    ///   - key: The experiment key to fetch.
    ///   - defaultValue: The default value to return if the experiment is not found or on error.
    ///                   This also determines the expected return type.
    ///   - cacheTTL: Cache time-to-live in milliseconds. Defaults to 24 hours.
    ///               Pass `nil` for no expiration, `0` to always fetch fresh.
    /// - Returns: The experiment value converted to the appropriate type, or the default value on error.
    ///
    /// Supported types:
    /// - `String` — maps to `STRING` (or `JSON`, returned raw)
    /// - `Int`, `Float`, `Double` — map to `NUMBER`
    /// - `Bool` — maps to `BOOLEAN`
    ///
    /// Example usage:
    /// ```swift
    /// let isFeatureEnabled = await manager.getExperiment(key: "new_feature", defaultValue: false)
    /// let maxRetries = await manager.getExperiment(key: "max_retries", defaultValue: 3)
    /// // Custom cache TTL (1 hour)
    /// let feature = await manager.getExperiment(key: "feature_flag", defaultValue: true, cacheTTL: 3_600_000)
    /// ```
    public func getExperiment<T>(key: String, defaultValue: T, cacheTTL: Int64? = 24 * 60 * 60 * 1000) async -> T {
        guard let result = await injector.experimentUseCase.get(key: key, cacheTTL: cacheTTL) else {
            return defaultValue
        }
        return convertExperimentValue(
            value: result.value,
            variableType: result.variableType,
            defaultValue: defaultValue
        )
    }

    /// Converts the experiment value string to the appropriate type based on the default value type.
    private func convertExperimentValue<T>(value: String, variableType: String, defaultValue: T) -> T {
        switch defaultValue {
        case is String:
            // A JSON experiment is handed back as its raw payload.
            guard variableType == "STRING" || variableType == "JSON" else { return defaultValue }
            return value as? T ?? defaultValue

        case is Bool:
            guard variableType == "BOOLEAN" else { return defaultValue }
            return (value.lowercased() == "true") as? T ?? defaultValue

        case is Int:
            guard variableType == "NUMBER", let intValue = Int(value) else { return defaultValue }
            return intValue as? T ?? defaultValue

        case is Float:
            guard variableType == "NUMBER", let floatValue = Float(value) else { return defaultValue }
            return floatValue as? T ?? defaultValue

        case is Double:
            guard variableType == "NUMBER", let doubleValue = Double(value) else { return defaultValue }
            return doubleValue as? T ?? defaultValue

        default:
            return defaultValue
        }
    }

    /// Destroys the `NativeblocksManager` and releases all associated resources.
    public func destroy() {
        for kit in kits {
            kit.detach(instanceName: name)
        }
        kits.removeAll()

        for loggerType in loggerTypes {
            removeLogger(instanceName: name, loggerType: loggerType)
        }
        loggerTypes.removeAll()

        NativeCoreSDKInjector.destroy(name: name)

        Self.lock.lock()
        Self.instanceRegistry.removeValue(forKey: name)
        Self.lock.unlock()

        NativeBlockProviderRegistry.remove(name)
        NativeActionProviderRegistry.remove(name)
        NativeTypeProviderRegistry.remove(name)
        actionContractors.removeAll()
    }
}
